File: backend-for-all/backend/src/main/java/com/nexaerp/role/RoleService.java
package com.nexaerp.role;

import com.nexaerp.role.dto.RoleRequestDto;
import com.nexaerp.role.dto.RoleResponseDto;

import java.util.List;

public interface RoleService {
    // Get all roles with their permissions
    List<RoleResponseDto> getAll();

    // Get single role by ID
    RoleResponseDto getById(Long id);

    // Create new role
    RoleResponseDto create(RoleRequestDto request);

    // Update role name, description, permissions
    RoleResponseDto update(Long id, RoleRequestDto request);

    // Assign permissions to role
    RoleResponseDto assignPermissions(Long id, java.util.Set<Long> permissionIds);

    // Remove permissions from role
    RoleResponseDto removePermissions(Long id, java.util.Set<Long> permissionIds);
}

File: accounting-system-web-project/nexa-erp-frontend/src/app/features/roles/services/role.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { APP_CONFIG } from '../../../core/config/app.config';
import { ApiResponse } from '../../../core/models/api-response.model';
import { Role } from '../models/role.model';

export interface RoleRequest {
  name: string;
  description: string;
  permissionIds: number[];
}

@Injectable({
  providedIn: 'root',
})
export class RoleService {
  private readonly baseUrl = `${APP_CONFIG.apiUrl}/roles`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<ApiResponse<Role[]>> {
    return this.http.get<ApiResponse<Role[]>>(this.baseUrl);
  }

  getById(id: number): Observable<ApiResponse<Role>> {
    return this.http.get<ApiResponse<Role>>(`${this.baseUrl}/${id}`);
  }

  create(request: RoleRequest): Observable<ApiResponse<Role>> {
    return this.http.post<ApiResponse<Role>>(this.baseUrl, request);
  }

  update(id: number, request: RoleRequest): Observable<ApiResponse<Role>> {
    return this.http.put<ApiResponse<Role>>(`${this.baseUrl}/${id}`, request);
  }

  assignPermissions(id: number, permissionIds: number[]): Observable<ApiResponse<Role>> {
    return this.http.post<ApiResponse<Role>>(
      `${this.baseUrl}/${id}/permissions/assign`,
      permissionIds,
    );
  }

  removePermissions(id: number, permissionIds: number[]): Observable<ApiResponse<Role>> {
    return this.http.post<ApiResponse<Role>>(
      `${this.baseUrl}/${id}/permissions/remove`,
      permissionIds,
    );
  }
}